src/components/Pagination/Pagination.jsx

```diff
@@ -1,163 +1,220 @@
+        pages: action.payload,
+      };
+    }
+    default:
+      return state;
+  }
+}
+
 const Pagination = props => {
   const theme = useTheme();
 
-  const [totalRecords, setTotalRecords] = useState(null);
-  const [totalPages, setTotalPages] = useState(0);
-  const [pageLimit, setPageLimit] = useState(10);
-  const [pageNeighbors, setPageNeighbors] = useState(1);
-  const [currentPage, setCurrentPage] = useState(1);
-  const [pages, setPages] = useState([]);
+  const [state, dispatch] = useReducer(reducer, {
+    totalRecords: null,
+    totalPages: 0,
+    pageLimit: 0,
+    pageNeighbors: 1,
+    currentPage: 1,
+    pages: [],
+  });
 
   useEffect(() => {
-    const { totalRecords, pageLimit, pageNeighbors } = props;
-
-    if (pageLimit) {
-      setPageLimit(pageLimit);
-    }
-    if (pageNeighbors) {
-      setPageNeighbors(Math.max(0, Math.min(pageNeighbors, 2)));
-    }
+    const {
+      currentPage = 1,
+      totalRecords = 0,
+      pageLimit = 10,
+      pageNeighbors = 1,
+    } = props;
 
-    const totalPages = Math.ceil(totalRecords / pageLimit);
-    setTotalRecords(totalRecords);
-    setTotalPages(totalPages);
-  }, []);
+    dispatch({
+      type: SET_PAGINATION_CONFIG,
+      payload: {
+        currentPage,
+        totalRecords,
+        pageLimit,
+        pageNeighbors: Math.max(0, Math.min(pageNeighbors, 2)),
+        totalPages: Math.ceil(totalRecords / pageLimit),
+      },
+    });
+  }, [
+    props.currentPage,
+    props.totalRecords,
+    props.pageLimit,
+    props.pageNeighbors,
+  ]);
 
   useEffect(() => {
-    setPages(fetchPageNumbers());
-  }, [totalPages, currentPage]);
-
-  const range = (from, to, step = 1) => {
-    let i = from;
-    const range = [];
-
-    while (i <= to) {
-      range.push(i);
-      i += step;
-    }
-
-    return range;
-  };
-
-  /**
-   * Let's say we have 10 pages and we set pageNeighbours to 2
-   * Given that the current page is 6
-   * The pagination control will look like the following:
-   *
-   * (1) < {4 5} [6] {7 8} > (10)
-   *
-   * (x) => terminal pages: first and last page(always visible)
-   * [x] => represents current page
-   * {...x} => represents page neighbours
-   */
-  const fetchPageNumbers = () => {
     /**
-     * totalNumbers: the total page numbers to show on the control
-     * totalBlocks: totalNumbers + 2 to cover for the left(<) and right(>) controls
+     * Let's say we have 10 pages and we set pageNeighbours to 2
+     * Given that the current page is 6
+     * The pagination control will look like the following:
+     *
+     * (1) < {4 5} [6] {7 8} > (10)
+     *
+     * (x) => terminal pages: first and last page(always visible)
+     * [x] => represents current page
+     * {...x} => represents page neighbours
      */
-    const totalNumbers = pageNeighbors * 2 + 3;
-    const totalBlocks = totalNumbers + 2;
+    const fetchPageNumbers = () => {
+      const { currentPage, pageNeighbors, totalPages } = state;
```

Comment:
Again, you don't want to do your destructuring within the `useEffect()` because it then sees all of `state` as a dependency.